clib/exninfo.ml

```diff
@@ -57,12 +57,29 @@ let rec find_and_remove_assoc (i : int) = function
     if rem == ans then (r, l)
     else (r, (j, v) :: ans)
 
-let iraise e =
+type backtrace = Printexc.raw_backtrace
+let backtrace_to_string = Printexc.raw_backtrace_to_string
+
+let backtrace_info : backtrace t = make ()
+
+let is_recording = ref false
+
+let record_backtrace b =
+  let () = Printexc.record_backtrace b in
+  is_recording := b
+
+let get_backtrace e = get e backtrace_info
+
+let iraise (e,i) =
   let () = Mutex.lock lock in
   let id = Thread.id (Thread.self ()) in
-  let () = current := (id, e) :: remove_assoc id !current in
+  let () = current := (id, (e,i)) :: remove_assoc id !current in
   let () = Mutex.unlock lock in
-  raise (fst e)
+  match get i backtrace_info with
+  | None ->
+    raise e
+  | Some bt ->
+    Printexc.raise_with_backtrace e bt
```

Comment:
This could be made conditional on `is_recording`. IMHO it is worth to raise_with_backtrace instead of what we did before.